Sfp readings

logicog opened the pull request:
<img width="1420" height="623" alt="Tooltips" src="https://github.com/user-attachments/assets/2c210a38-a3c4-425e-ba37-83feacca1f0d" />

This pull request adds support for displaying of SFP dynamic information such as the temperature as tooltips for SFP ports. For Ethernet ports the speed is displayed, but it would also be possible to e.g. add link state information such as advertised speeds.

A simulation of SFP data is added for http_sim.

Because SFF-8472 requires bulk reading of the 2-byte dynamic data such as the module temperature, bulk reading of up to 16 bytes of SFP-module data via I2C bulk transfers is added. This could also be used to bulk transfer e.g. the module name.

Display of the raw SFP module data is added to the CLI. However, no pretty printing of the values e.g. in mW is done because of the requirement for custom tailored conversion routines without the availability of floating-point numbers.

vDorst commented on html/main.js:
```diff
@@ -72,6 +83,15 @@ function update() {
 	  }
+	  tt.innerHTML = "Link speed: " + linkS[p.link + 1] + "<br>";
+	  if (p.isSFP) {
+	    tt.innerHTML += "T = " + (Number(p.sfp_temp) >> 8) + "." + ((Number(p.sfp_temp) & 0xff)/256.0 * 100).toFixed(0) + "&deg;C<br>";
```
instead of `&deg;C`, `&#8451;` can be used as "Degree Celsius".

logicog replied:
I'll try that, it might look nicer on the page.

vDorst commented on rtlplayground.c:
```diff
@@ -745,6 +745,15 @@ void sds_config(uint8_t sds, uint8_t mode)
+		reg &= 0x7f;
+	} else {
+		// Configure SFP EEPROM address (0x50) as I2C device address
+		REG_SET(RTL837X_REG_I2C_CTRL, 0x1L << I2C_MEM_ADDR_WIDTH | 0x50 << I2C_DEV_ADDR);
```
Compiler warnings!
```
rtlplayground.c:750: warning 116: left shifting more than size of object changed to zero
-:0: warning 116: left shifting more than size of object changed to zero
rtlplayground.c:754: warning 116: left shifting more than size of object changed to zero
```

logicog commented on httpd/page_impl.c:
```diff
@@ -148,6 +148,46 @@ void send_sfp_info(uint8_t sfp)
+	} else {
+		REG_WRITE(RTL837X_REG_I2C_CTRL, 0x00, 0x1 << (I2C_MEM_ADDR_WIDTH-16) | len & 0xf,  0x50 >> 5, 0x50 << 3);
+	}
+	reg &= 0x7f;
```
I'll do the first as it is more readable.

vDorst replied:
This was more about to show what I tried to get better machine-code output.
Not really a change suggestion

logicog commented on httpd/httpd.c:
```diff
@@ -593,7 +593,7 @@ void httpd_appcall(void)
 			slen = strtox(outbuf, "HTTP/1.1 200 OK\r\nContent-Type: ");
 			slen += strtox(outbuf + slen, mime_strings[f_data[entry].mime]);
-			slen += strtox(outbuf + slen, "\r\nCache-Control: max-age=60, must-revalidate\r\n\r\n");
+			slen += strtox(outbuf + slen, "; charset=UTF-8\r\nCache-Control: max-age=60, must-revalidate\r\n\r\n");
```
Yes. Without this, the Degree sign does not work, instead funny characters appear. I thought that the charset directive configures the data being transmitted, but it appears more that it defines what characters can be displayed. UTF-8 is anyway the default nowadays, and we may need this to support translated pages with other European and Asian languages. Note that the charset directive is only for pages served out of the flash. json data is not affected.

logicog commented on html/main.js:
```diff
@@ -72,6 +83,15 @@ function update() {
+	    tt.innerHTML += "Vcc = " + (Number(p.sfp_vcc) / 10000.0).toFixed(2) + "V<br>";
+	    tt.innerHTML += "TX-Bias = " + (Number(p.sfp_txbias) / 500.0).toFixed(1) + "mA<br>";
+	    tt.innerHTML += "TX-Power = " + (Number(p.sfp_txpower) / 10.0).toFixed(0) + "mW<br>";
+	    tt.innerHTML += "RX-Power = " + (Number(p.sfp_rxpower) / 10.0).toFixed(0) + "mW<br>";
```
I will try  a narrow no-break space : &#8239;

logicog commented:
I implemented all of  your suggestions.
I also added tabular design for the tool-tips. However, I am not a web-designer and have to pour a lot of time into something that afterwards only looks mediocre. At least, now it is a table, and the rest can be done in style.css.

logicog commented:
Fixed all of the above and rebased.
